otelx/semconv/context.go

```diff
@@ -0,0 +1,47 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package semconv
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/urfave/negroni"
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/ory/x/httpx"
+)
+
+type contextKey int
+
+const contextKeyAttributes contextKey = iota
+
+func ContextWithAttributes(ctx context.Context, attrs ...attribute.KeyValue) context.Context {
+	existing, _ := ctx.Value(contextKeyAttributes).([]attribute.KeyValue)
+	return context.WithValue(ctx, contextKeyAttributes, append(existing, attrs...))
+}
+
+func AttributesFromContext(ctx context.Context) []attribute.KeyValue {
+	fromCtx, _ := ctx.Value(contextKeyAttributes).([]attribute.KeyValue)
+	uniq := make(map[attribute.Key]struct{})
+	attrs := make([]attribute.KeyValue, 0)
+	for i := len(fromCtx) - 1; i >= 0; i-- {
+		if _, ok := uniq[fromCtx[i].Key]; !ok {
+			uniq[fromCtx[i].Key] = struct{}{}
+			attrs = append(attrs, fromCtx[i])
+		}
+	}
+	reverse(attrs)
+	return attrs
+}
+
+var Middleware negroni.HandlerFunc = func(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	next(rw, r.WithContext(ContextWithAttributes(r.Context(), AttrClientIP(httpx.ClientIP(r)))))
+}
```

Comment:
Is there a reason not to write this as a function?
```suggestion
func Middleware(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
	next(rw, r.WithContext(ContextWithAttributes(r.Context(), AttrClientIP(httpx.ClientIP(r)))))
}
```

Reply:
Yep, that's a useful suggestion. It would decouple this package from urfave/negroni (somewhat).